[2, 3단계 - 지하철 노선도 미션] 체프(심문성) 제출합니다.

Puterism opened the pull request:
## [🚀 데모 페이지](https://wooteco-subway-cheffe.herokuapp.com/)

안녕하세요! 파노(@jho2301)와 함께 페어 프로그래밍을 진행한 체프입니다. 
이번 리뷰도 잘 부탁드립니다!

데모 페이지의 테스트 계정은 다음과 같습니다. (다른 계정으로 회원가입하고 접속하셔도 됩니다!)
```
ID: test@test.com
PW: test
```

### [⬅️ 1단계 PR](https://github.com/woowacourse/javascript-subway/pull/4)
추가로 1단계 PR에 남겨주신 리뷰에도 답글을 달아뒀는데 확인 부탁드릴게요 :D


## 시연

- 지하철 역을 추가, 수정, 삭제할 수 있다.
![subway_station1](https://user-images.githubusercontent.com/2542730/113466680-a202a200-9478-11eb-89b7-18a6e0df6734.gif)

- 이미 등록된 노선의 경우 삭제할 수 없다.
![subway_station3](https://user-images.githubusercontent.com/2542730/113466682-a29b3880-9478-11eb-8574-55dc7a9bfeb2.gif)

- 상행역과 하행역 정보, 그리고 노선 이름과 색상을 입력하여 노선을 추가할 수 있다.
![subway_line1](https://user-images.githubusercontent.com/2542730/113466675-9ca55780-9478-11eb-99fc-7c2fb8a66404.gif)

- 노선 정보를 수정할 수 있다.
![subway_line2](https://user-images.githubusercontent.com/2542730/113466676-9e6f1b00-9478-11eb-8e91-0f6155d33316.gif)

- 각 노선별로 구간을 추가할 수 있다. 구간을 추가할 때의 역은 추가 가능한 역의 목록에서 선택할 수 있다.
![subway_sections](https://user-images.githubusercontent.com/2542730/113466678-a0d17500-9478-11eb-9afb-5fd5f7a70c73.gif)

- 전체 노선을 조회할 수 있다. 환승 가능한 역도 함께 보여지며, 클릭할 경우에 해당 노선에서의 역에 하이라이트된다.
![subway_map](https://user-images.githubusercontent.com/2542730/113466677-9fa04800-9478-11eb-8e85-6c53943c1661.gif)

jho2301 committed:
docs: step2 기능구현목록 업데이트

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 committed:
refactor: 기본 main ui 삭제

jho2301 committed:
chore: API BASE_URL 변경

jho2301 committed:
test: 지하철 역을 등록할 수 있다

jho2301 committed:
feat: 지하철역을 등록할 수 있다.

- 지하철역은 2글자 이상이어야 한다
- 엔터키 또는 마우스 클릭으로 역을 추가할 수 있어야 한다.
- 지하철역의 이름은 최대 20자 이하이다.

jho2301 committed:
feat: api에러 시 error 페이지 구현

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 committed:
feat: api호출 시 loading 페이지 구현

jho2301 committed:
feat: 지하철 역 리스트 조회 기능

jho2301 committed:
feat: 중복된 지하철 역은 추가할 수 없다.

jho2301 committed:
feat: authenticatedRoute strict 모드 추가

- strict 모드: api호출을 통해 로그인중인지 확인
- sloppy 모드: 프로세스 내 상태값에 access token이 있는지 확인

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 committed:
test: 지하철역을 삭제할 수 있다

jho2301 committed:
chore: cypress defaultCommandTimeout 옵션 변경

8000 -> 3000

jho2301 committed:
feat: 지하철역을 삭제할 수 있다

jho2301 committed:
feat: 삭제 시 confirm을 이용하여 한 번 더 유저가 확인할 수 있어야 한다.

jho2301 committed:
test: 지하철역을 수정할 수 있다.

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 committed:
feat: 지하철역의 이름을 수정할 수 있다.

- 수정 ui는 직접 구현한다.

jho2301 committed:
fix: 역삭제, 역수정이 UI에 반영되지 않는 현상

jho2301 committed:
feat: 역 이름 수정 프로세스, 모달로 구현

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 committed:
refactor: constants.js 도메인 별로 분리

- 노선 추가 모달 open 기능구현

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 committed:
feat: 노선 추가시, 노선 색을 선택하는 기능 (초기값 랜덤)

jho2301 committed:
feat: 노선 등록시, 최소 역 개수를 충족 해야한다.

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 committed:
feat: 노선 조회, 삭제, 생성 기능

jho2301 committed:
refactor: store 객체들을 감싸는 store 객체 추가

- store객체임을 명시적 표현

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 committed:
fix: 로그인 실패시, 상태가 초기화되는 버그

jho2301 committed:
fix: 노선 추가시, 이전 form input값들이 제거되지 않는 현상

jho2301 committed:
refactor: viewController디렉토리 도메인별로 재구성

Puterism committed:
feat: 지하철 노선 이름과 색깔을 수정하는 기능

Co-authored-by: hwano <jho2301@gmail.com>

jho2301 committed:
refactor: 디렉토리 재구성

- 함수 기능별 디렉토리구분에서 page별 디렉토리 구조로 변경

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 committed:
feat: esc 키로 모달 닫기 기능

jho2301 committed:
fix: 서버에서 전달하는 에러메세지로 alert메시지를 표시하는 기능

jho2301 committed:
fix: 지하철역 삭제시, store에 반영되지 않는 현상

jho2301 committed:
test:  테스트케이스 추가

- 노선을 등록할 수 있다.
- 노선을 수정할 수 있다.
- 노선을 삭제할 수 있다.

jho2301 committed:
feat: 구간을 조회할 수 있다.

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 committed:
feat: 지하철 노선의 구간을 조회할 수 있다.

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 committed:
feat: 회원가입 페이지에서 로그인하러 가기 기능 추가

jho2301 committed:
chore: semver 0.0.1로 변경

jho2301 committed:
refactor: 공통 element tag로 핸들링한 스타일시트를 classname으로 변경

jho2301 committed:
refactor: 코드 리팩토링

- handleLogin 파라미터에 구조 분해 할당 적용
- colorOptions객체에 Object.freeze() 적용
- timerId에 undefined → null 넣기

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 committed:
feat: 구간 추가 기능

- services/ 의 api요청 함수들을 api/ 로 이동

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 committed:
feat: 구간 삭제 기능

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 committed:
feat: 구간 삭제 기능

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 committed:
feat: 이미 노선에 등록된 역인 경우 삭제할 수 없다

jho2301 committed:
fix: 구간 관리에서 버튼 이외 영역 클릭 시, 얼리리턴 되지 않는 현상

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 committed:
feat: 상행 종점 추가 기능

jho2301 committed:
fix: 노선 추가 모달 select태그 드롭다운아이콘 표시되지 않는 현상

jho2301 committed:
fix: UI 예외상황 처리

- 구간관리 종점에 추가 버튼 중앙정렬하기
- 노선관리 노선추가 모달에 컬러픽커 가운데 정렬

jho2301 committed:
fix: input 및 button에 폰트 적용되지 않는 현상

jho2301 committed:
fix: 여러 번 구간 추가시, 거리, 시간입력폼의  max속성이 초기화되지 않는 현상

jho2301 committed:
test: 테스트 코드 추가

- 지하철 구간 관리
- 지하철 노선 관리

jho2301 committed:
test(fix): 의도대로 되지않는 테스트 동작 수정

- after함수에서 역이 삭제되지 않는 현상
- 수행한 동작과 검사동작 일치시킴
- 잘못 사용된 cy command 수정

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 committed:
fix: 노선 삭제 후 상태에 반영이 되지 않는 현상

jho2301 committed:
fix: 구간 조회 시 구간의 순서가 보장되지 않는 현상

jho2301 committed:
fix: 구간 삭제시, 상태 업데이트 이전에 렌더되는 현상

Puterism committed:
docs: Step3 기능 목록 업데이트

Puterism committed:
chore: 더 이상 사용하지 않는 메뉴 및 코드 삭제

Puterism committed:
ui: 전체보기 기능 UI 구현

Puterism committed:
feat: 노선별로 전체 보기 기능 구현

Puterism committed:
feat: 전체보기 기능에서 역에 환승 가능한 노선을 표시한다.

- 환승 가능한 노선 버튼을 클릭하면, 해당 역을 하이라이트하는 기능

Puterism committed:
chore: webpack 설정 변경
- output.filename에서 chunkhash에서 contenthash로 변경

Puterism committed:
docs: README.md 오타 수정

Puterism committed:
feat: alert 창을 snackbar로 변경
- 노선, 역, 구간이 추가되거나 수정될 때도 snackbar message가 보이도록 구현

Puterism committed:
fix: 환승역이 하나일 때 제대로 출력되지 않았던 문제 수정
- 환승역 이미지 사이즈 줄여서 적용
- 의도하지 않게 변경된 webpack 설정 복원

Puterism committed:
Merge branch 'puterism' into cheffe-step3

Vallista commented on cypress/integration/subway.spec.js:
```diff
@@ -1,6 +1,77 @@
-describe('지하철 노선도', () => {
+const id = 'test123@test123.com';
```
id와 pw는 추후 노출이 될 수도 있으니, json으로 만들어서 dynamic import (require)를 통해 파일을 불러오고, 해당 계정 파일은 실행 시점에 생성해주고, ignore 시키는 처리가 있으면 좋겠습니다.

Puterism replied:
앗, id와 pw에 대한 처리를 해주지 않았었군요... ㅠㅠ 짚어주셔서 감사드립니다.

별도의 json 파일로 관리해주기 위해 관련 정보를 검색하다가 Cypress fixture 기능에 대해서 알게 되었습니다. Cypress의 fixture 폴더에 테스트 계정의 id, pw 정보를 담은 json을 생성하여 사용하는 방식으로 변경하였습니다.

Puterism replied:
> 해당 계정 파일은 실행 시점에 생성해주고, ignore 시키는 처리가 있으면 좋겠습니다.

말씀해주신 부분에서 위 문장이 잘 이해되지 않는데, 테스트를 위해 사용되는 계정을 테스트 실행 시점에 임의로 회원가입했다가, 테스트 종료 후 계정을 서버에서 삭제되도록 처리하는 방식을 제안하신 것으로 이해했는데, 제가 맞게 이해한 걸까요? 🤔

Vallista replied:
```
앗, id와 pw에 대한 처리를 해주지 않았었군요... ㅠㅠ 짚어주셔서 감사드립니다.

별도의 json 파일로 관리해주기 위해 관련 정보를 검색하다가 Cypress fixture 기능에 대해서 알게 되었습니다. Cypress의 fixture 폴더에 테스트 계정의 id, pw 정보를 담은 json을 생성하여 사용하는 방식으로 변경하였습니다.
```

요 부분을 보면 이해 하신 것 같습니다 :)

Vallista commented on cypress/integration/subway.spec.js:
```diff
@@ -1,6 +1,77 @@
+  });
+}
+
+describe('지하철 노선도 STEP1', () => {
```
STEP1 처럼 describe의 이름을 정하기보다, 명확한 컨텍스트를 적으면 좋을 것 같습니다.

Puterism replied:
다른 describe와 문맥이 맞도록 어떤 테스트를 진행하는 것인지에 대해서 설명하는 네이밍으로 변경해두겠습니다! 감사합니다!

Vallista commented on cypress/integration/subway.spec.js:
```diff
@@ -9,3 +80,133 @@ describe('지하철 노선도', () => {
   });
 });
+
+describe('지하철 역 관리', () => {
```
describe마다 파일을 나누는 건 어떨까요?

Puterism replied:
저도 파일을 나누는 게 좋다고 생각합니다!
처음에는 파일을 나눠서 테스트 코드를 작성하면, 각 테스트 코드 파일 별로만 테스트를 실행할 수밖에 없는 줄 알았는데... 모든 테스트 파일을 한번에 테스트 하는 방법도 있었네요 😅

Vallista commented on cypress/integration/subway.spec.js:
```diff
@@ -9,3 +80,133 @@ describe('지하철 노선도', () => {
+      downStation: '연세대',
+      duration: 10,
+      distance: 10,
+      color: 'bg-gray-300',
```
color 값은 constant로 관리하면 좋겠네여

Puterism replied:
color 값이 정해져 있어서 저도 constant로 관리하면 더 좋을 것 같습니다. 피드백 감사합니다!

Vallista commented on cypress/integration/subway.spec.js:
```diff
@@ -9,3 +80,133 @@ describe('지하철 노선도', () => {
+      color: 'bg-gray-300',
+    });
+
+    cy.get('.js-line-list-item[data-name="로이드호선호"]').should('exist');
```
`.js-line-list-item[data-name="로이드호선호"]` 요런 형태도 constant로 관리할 수 있을 것 같아욤

Puterism replied:
이 부분은 data-name 을 입력받는 함수의 형태로 관리하면 좋을 것 같습니다. 피드백 감사드립니다!

Vallista commented on index.html:
```diff
@@ -9,7 +9,7 @@
       <div class="d-flex justify-center mt-5 w-100">
         <div class="w-100">
-          <header id="header" class="my-4">
+          <header id="header" class="my-4" style="z-index: 10">
```
css가 inline style로 있다면.. 여러군데 파편화 되어 있으면 관리하기 어려우니 css에서 관리를 하는게 더 좋지 않을까 합니다.

Puterism replied:
앗... 스타일로 옮겨두려고 했는데 깜빡했습니다 ㅠㅠ 짚어주셔서 감사합니다!

Vallista commented on src/css/index.css:
```diff
@@ -63,6 +76,98 @@ ul li button:hover {
+.list-item:first-child .section-edit,
+.list-item:last-child .section-edit {
+  transform: translateX(50%);
+}
```
eof를 지키면 좋겠습니다 :)

Puterism replied:
헉... 지적 감사드립니다!

Vallista commented on src/css/pages/map.css:
```diff
@@ -0,0 +1,118 @@
+    background-color: yellow;
+    opacity: 0;
+  }
+}
```
eof!

Puterism replied:
⭐EOF⭐

Vallista commented on src/css/ui/modules/spinner.css:
```diff
@@ -0,0 +1,30 @@
+  animation-iteration-count: infinite;
+}
+
+@keyframes spinnerAnimation {
```
`@keyframes spinner`가 오히려 맞지 않을까 싶어요~ keyframes는 animation에 쓰이니

Puterism replied:
spinner 라는 이름만으로는 keyframes 임을 알 수 없을 거라고 생각해서 뒤에 `Animation` 을 붙였었는데, 어차피 keyframes는 animation에서만 사용되니까 굳이 붙여둘 필요는 없었겠네요 😅 감사합니다!

Vallista commented on src/js/api/auth.js:
```diff
@@ -0,0 +1,67 @@
+  } catch (error) {
+    return {
+      success: false,
+      message: error.message ?? AUTH.SIGNUP_FAILED,
```
error.message ?? 는 message가 ''일 때는 체크를 못해요. 그러므로 error.message || AUTH.SIGNUP_FAILED가 맞지않을까 싶어요.

Puterism replied:
확실히 `||`가 커버할 수 있는 범위가 더 많아보이네요! 같이 수정해두겠습니다! 감사합니다~

Vallista commented on src/js/api/line.js:
```diff
@@ -0,0 +1,88 @@
+  } catch (error) {
+    return {
+      success: false,
+      message: error.message ?? LINE.GET_LINE_LIST_FAILED,
```
요기도 위와 같습니당

Vallista commented on src/js/api/line.js:
```diff
@@ -0,0 +1,88 @@
+  } catch (error) {
+    return {
+      success: false,
+      message: error.message ?? LINE.ADD_LINE_FAILED,
```
요기두~

Vallista commented on src/js/pages/lines/eventHandlers/handleLineStatus.js:
```diff
@@ -0,0 +1,43 @@
+    openModal($('#line-edit-modal'));
+    $('#line-edit-modal #subway-line-name').focus();
+
+    return;
```
오 여긴 return이 없어도 될 것 같네요!

Puterism replied:
엇, 불필요한 return 이었네요! 짚어주셔서 감사드립니다~

Vallista reviewed (changes requested):
안녕하세요 체프님!
2, 3단계 미션 고생 많으셨습니다!

1단계 코맨트에 추가하여 남겨두었습니다 👍 
2,3단계 관련하여 코드라인에 남겨두었습니다 👍 

레이아웃 적으로는 별도의 피드백은 따로 없었습니다. 전반적인 코드 피드백만 드리도록 하겠습니다!

## 전반적인 코드 피드백

1. 앞서 피드백에 드린대로 constant로 class, id로 관리를 하는 것이 안정성 vs 효율성 이라고 피드백을 드렸는데요~ class, id를 빈번하게 불러오는 상황에서 어떤 방법이 더 장기적으로 좋을지 고민을 해보면 좋을 것 같습니다.

2. 에러가 발생하는 상황을 정의해보고, 해당 상황에 대해서 쉽게 알 수 있도록 하는 일련의 무언가가 필요해보입니다. parameter에 의존적인 함수의 경우, parameter가 제대로 들어오지 않았다면 어떤 값으로 들어왔는지, 의도한 값인지에 대한 테스트 혹은 예외처리등이 필요하겠지요. 이러한 부분에 대해서 조금 더 고민을 해보면 좋을 것 같습니다. (또한 API 레벨도 동일합니다.)

3. 그 외 디테일한 부분은 코드 라인에 적어두었습니다 👍  전반적으로 깔끔한 코드라 피드백이 많이 없었네요.. (흑)

Puterism committed:
test(refactor): 코드 리뷰에 따른 리팩토링
- 컨텍스트를 더 명확하도록 describe의 이름 다시 정의
- login 에 사용되는 email, password를 별도의 json fixture로 관리
- 페이지 동작 관련 함수들을 support 폴더에 별도로 분리

Puterism committed:
test(refactor): describe 별 테스트 코드 분리

Puterism committed:
refactor: 코드 리뷰에 따른 리팩토링
- inline으로 적용된 style을 index.css에 옮겨서 정의
- EOF가 지켜지지 않은 파일 수정
- keyframes 이름의 네이밍 변경

Puterism committed:
refactor: 불필요한 return문 삭제

Puterism committed:
fix: api 함수에서 `error.message`가 빈 문자열일 경우를 올바르게 처리하지 않는 문제 수정

Puterism committed:
refactor: 노선 색상 상수화 및 테스트 코드에 적용

Puterism commented:
> 앞서 피드백에 드린대로 constant로 class, id로 관리를 하는 것이 안정성 vs 효율성 이라고 피드백을 드렸는데요~ class, id를 빈번하게 불러오는 상황에서 어떤 방법이 더 장기적으로 좋을지 고민을 해보면 좋을 것 같습니다.

장기적으로 보았을 때, 좀 더 안정적으로 class와 id 명을 관리해야 한다면 상수로 만드는 것이 더 좋다고 생각합니다.
만약 class와 id 명을 상수로 관리해야한다면 정말 많은 값들을 관리해야 하고, 어느 정도 컨벤션도 존재해야 할 것 같습니다.
좀 더 효율적으로 관리할 방법에 대해서 더 고민해봐야 할 것 같다는 생각이 들었습니다.

지금까지는

- id와 class를 구별하여 관리한다.
- 파일이 커질 경우 도메인 별로 분리한다.
- `#`이나 `.` 문자와를 함께 사용하는 selector의 형태로 사용하지 않는 경우도 있으니, 순수하게 이름만 저장한다.

정도가 떠오릅니다.

> 에러가 발생하는 상황을 정의해보고, 해당 상황에 대해서 쉽게 알 수 있도록 하는 일련의 무언가가 필요해보입니다. parameter에 의존적인 함수의 경우, parameter가 제대로 들어오지 않았다면 어떤 값으로 들어왔는지, 의도한 값인지에 대한 테스트 혹은 예외처리등이 필요하겠지요. 이러한 부분에 대해서 조금 더 고민을 해보면 좋을 것 같습니다. (또한 API 레벨도 동일합니다.)

확실히 지금 상태에서는 함수 사용 시의 예외 처리가 많이 되어 있지 않은 것 같습니다.
예를 들어, parameter를 객체의 형태로 받는 경우에, 기본값 정의가 되어있은 경우가 많아, undefined된 값으로 바로 함수의 로직이 실행될 여지가 보였습니다.

예외 처리를 위해서 다음 사항을 고려해봐야 할 것 같습니다.

- 함수 parameter의 기본값을 지정해준다.
- 만약 꼭 입력 받아야 하는 parameter를 입력받지 못했을 경우, error를 throw 한다.
- 예외 처리를 해주어야 하는 함수가 많고, 코드 라인이 길어져 함수 로직과 에러 핸들링 로직이 구분되지 않을 정도로 복잡해진다면, 에러 핸들링의 영역을 분리한다. 예를 들어서 errorHandler 와 같은 객체나 함수를 만들어서 적용하는 방법이 있을 것이다.


피드백 감사드립니다!

Vallista commented:
감사합니다 체프님!

```
장기적으로 보았을 때, 좀 더 안정적으로 class와 id 명을 관리해야 한다면 상수로 만드는 것이 더 좋다고 생각합니다.
만약 class와 id 명을 상수로 관리해야한다면 정말 많은 값들을 관리해야 하고, 어느 정도 컨벤션도 존재해야 할 것 같습니다.
좀 더 효율적으로 관리할 방법에 대해서 더 고민해봐야 할 것 같다는 생각이 들었습니다.

지금까지는

id와 class를 구별하여 관리한다.
파일이 커질 경우 도메인 별로 분리한다.
#이나 . 문자와를 함께 사용하는 selector의 형태로 사용하지 않는 경우도 있으니, 순수하게 이름만 저장한다.
정도가 떠오릅니다.
```

컨벤션은 팀마다 각자 존재하고, 본인만의 컨벤션도 존재하니까 자기만의 무언가를 만들어두는 것은 좋은 경험일 거라 생각합니다 :)

```
확실히 지금 상태에서는 함수 사용 시의 예외 처리가 많이 되어 있지 않은 것 같습니다.
예를 들어, parameter를 객체의 형태로 받는 경우에, 기본값 정의가 되어있은 경우가 많아, undefined된 값으로 바로 함수의 로직이 실행될 여지가 보였습니다.

예외 처리를 위해서 다음 사항을 고려해봐야 할 것 같습니다.

함수 parameter의 기본값을 지정해준다.
만약 꼭 입력 받아야 하는 parameter를 입력받지 못했을 경우, error를 throw 한다.
예외 처리를 해주어야 하는 함수가 많고, 코드 라인이 길어져 함수 로직과 에러 핸들링 로직이 구분되지 않을 정도로 복잡해진다면, 에러 핸들링의 영역을 분리한다. 예를 들어서 errorHandler 와 같은 객체나 함수를 만들어서 적용하는 방법이 있을 것이다.
```

해당 고려한 부분을 사실 시간에 맞게 짜기 어렵긴 하지만... 그렇다면 시간을 잘 맞추면서 해결할 수 있는 방법은 무엇인지. 생각해봐도 좋을 것 같습니다. (저는 개인적으로 TDD 선호)

Vallista commented:
피드백에 대해서 이해와, OK를 해주셨으니 머지하도록 하겠습니다! 추후 미션에서 더 멋진 활약 기대할께요!
